src/coreclr/nativeaot/Runtime/wasm/PalRedhawkWasm.cpp

```diff
@@ -39,4 +39,101 @@ void PalGetMaximumStackBounds_SingleThreadedWasm(void** ppStackLowOut, void** pp
     *ppStackLowOut = pStackLow;
     *ppStackHighOut = pStackHigh;
 }
+
+#ifdef TARGET_WASI
+// nop stubs
+
+int pthread_mutex_init(pthread_mutex_t * m, const pthread_mutexattr_t *__restrict a)
+{
+    return 0;
+}
+
+int pthread_mutexattr_init(pthread_mutexattr_t *a)
+{
+    return 0;
+}
+
+int pthread_mutexattr_settype(pthread_mutexattr_t *a, int type)
+{
+    return 0;
+}
+
+int pthread_mutex_destroy(pthread_mutex_t *mutex)
+{
+    return 0;
+}
+
+int pthread_mutexattr_destroy(pthread_mutexattr_t *)
+{
+    return 0;
+}
+
+int pthread_cond_init(pthread_cond_t *cond, const pthread_condattr_t *attr)
+{
+    return 0;
+}
+
+int pthread_condattr_init(pthread_condattr_t *attr)
+{
+    return 0;
+}
+
+int pthread_mutex_lock(pthread_mutex_t *)
+{
+    return 0;
+}
+
+int pthread_mutex_unlock(pthread_mutex_t *)
+{
+    return 0;
+}
+
+pthread_t pthread_self(void)
+{
+    return (pthread_t)0;
+}
+
+int pthread_equal(pthread_t, pthread_t)
+{
+    return 1; // only one thread
+}
+
+int pthread_attr_init(pthread_attr_t *)
+{
+    return 0;
+}
+
+int pthread_attr_destroy(pthread_attr_t *)
+{
+    return 0;
+}
+
+int pthread_condattr_destroy(pthread_condattr_t *)
+{
+    return 0;
+}
+
+int pthread_cond_broadcast(pthread_cond_t *)
+{
+    return 0;
+}
+
+int pthread_attr_setdetachstate(pthread_attr_t *, int)
+{
+    return 0;
+}
+
+#include <sys/mman.h>
+int mprotect(void* addr, size_t len, int prot)
+{
+    return 0;
+}
+
+extern "C" int __cxa_thread_atexit(void (*func)(), void*, void*)
+{
+    return 0;
+}
```

Comment:
This is a bit interesting... Who is calling this?

Reply:
```
error: failed to run `E:\GitHub\runtimelab\artifacts\tests\coreclr\wasi.wasm.Debug\nativeaot\SmokeTests\HelloWasm\HelloWasm\native\HelloWasm.wasm`
╰─▶ 1: RuntimeError: unreachable
           at undefined_stub (HelloWasm.wasm[29]:0xf63f)
           at __cxx_global_var_init (HelloWasm.wasm[484]:0x30de9)
           at thread-local initialization routine for tls_destructionMonitor (HelloWasm.wasm[558]:0x33d44)
           at thread-local wrapper routine for tls_destructionMonitor (HelloWasm.wasm[488]:0x30f2f)
           at PalAttachThread (HelloWasm.wasm[487]:0x30ef3)
           at ThreadStore::AttachCurrentThread(bool) (HelloWasm.wasm[368]:0x22c8b)
           at ThreadStore::AttachCurrentThread() (HelloWasm.wasm[376]:0x23156)
           at Thread::ReversePInvokeAttachOrTrapThread(ReversePInvokeFrame*) (HelloWasm.wasm[225]:0x1d437)
           at RhpReversePInvokeAttachOrTrapThread2 (HelloWasm.wasm[234]:0x1d9bc)
           at RhpReversePInvoke (HelloWasm.wasm[235]:0x1dc24)
           at S_P_CoreLib_Internal_Runtime_CompilerHelpers_StartupCodeHelpers__InitializeModules (HelloWasm.wasm[7101]:0x687c96)
           at InitializeRuntime() (HelloWasm.wasm[24025]:0x10b5b0e)
           at main (HelloWasm.wasm[24024]:0x10b5a2f)
           at __main_void (HelloWasm.wasm[24027]:0x10b5bbb)
           at _start (HelloWasm.wasm[39]:0xf689)
```

Reply:
I couldn't find the source for `__cxx_global_var_init` but without this stub, you get this error

Reply:
This looks like a toolchain bug, it shouldn't be generating calls to undefined functions (i. e. `atext` when threading is disabled).

We will need a functional implementation until it's fixed (something in the spirit of https://github.com/llvm-mirror/libcxxabi/blob/master/src/cxa_thread_atexit.cpp, but for a single thread, I assume)